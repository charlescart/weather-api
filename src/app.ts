/* eslint-disable no-console */
import app from './server';

const port = process.env.PORT || 3000;

app.listen(port, () => {
  console.log(`Run in the port ${port}`);
});
