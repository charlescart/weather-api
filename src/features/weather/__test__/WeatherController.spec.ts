import app from '../../../server';
import request from 'supertest'

const toMatch = {
  infoFromYourIp: {
    coord: {
      lat: expect.any(Number),
      lon: expect.any(Number),
    },
  },
  weather: {
    coord: {
      lat: expect.any(Number),
      lon: expect.any(Number),
    },
    forescast: expect.any(Object),
  },
};

describe('WeatherController', () => {
  describe('EndPoints Location, Current and Forecast', () => {
    it('GET /v1/location with IP from San Francisco/US', async (done) => {

      return request(app).get('/v1/location')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '162.243.152.177')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot(toMatch);
          done();
        });
    }, 30000);

    it('GET /v1/location with IP invalid ::1 in localhost', async (done) => {

      return request(app).get('/v1/location')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '::1')
        .expect(400)
        .then((res) => {
          expect(res.body).toMatchSnapshot();
          done();
        });
    }, 30000);

    it('GET /v1/current without city name and with IP from Hurlingham/Argentina', async (done) => {

      return request(app).get('/v1/current')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: expect.any(Object),
            temp: expect.any(Number),
            tempMax: expect.any(Number),
            tempMin: expect.any(Number),
            feelsLike: expect.any(Number),
            description: expect.any(String),
          });

          done();
        });
    }, 30000);

    it('GET /v1/current without city name and with IP from Bengaluru/India', async (done) => {

      return request(app).get('/v1/current')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '142.93.217.248')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: expect.any(Object),
            temp: expect.any(Number),
            tempMax: expect.any(Number),
            tempMin: expect.any(Number),
            feelsLike: expect.any(Number),
            description: expect.any(String),
          });

          done();
        });
    }, 30000);

    it('GET /v1/current/caracas with city name Caracas/Venezuela :D :D :D', async (done) => {

      return request(app).get('/v1/current/caracas')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: expect.any(Object),
            temp: expect.any(Number),
            tempMax: expect.any(Number),
            tempMin: expect.any(Number),
            feelsLike: expect.any(Number),
            description: expect.any(String),
          });

          done();
        });
    }, 30000);

    it('GET /v1/current/127barcelona52es7329437 with city name impossible', async (done) => {

      return request(app).get('/v1/current/127barcelona52es7329437')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: expect.any(Object),
            temp: expect.any(Number),
            tempMax: expect.any(Number),
            tempMin: expect.any(Number),
            feelsLike: expect.any(Number),
            description: expect.any(String),
          });

          done();
        });
    }, 30000);

    it('GET /v1/current/h with city name very short', async (done) => {

      return request(app).get('/v1/current/h')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(400)
        .then((res) => {
          expect(res.body).toMatchSnapshot();

          done();
        });
    }, 30000);

    it('GET /v1/forecast without city name and with IP from Toronto/Canada', async (done) => {

      return request(app).get('/v1/forecast')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '142.93.148.182')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: {
              lat: expect.any(Number),
              lon: expect.any(Number),
            },
            forescast: expect.any(Object)
          });

          done();
        });
    }, 30000);

    it('GET /v1/forecast/japan with city name Japan/Japan', async (done) => {

      return request(app).get('/v1/forecast/japan')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: {
              lat: expect.any(Number),
              lon: expect.any(Number),
            },
            forescast: expect.any(Object)
          });

          done();
        });
    }, 30000);

    it('GET /v1/forecast/&1$7japan?*,jp with city name impossible', async (done) => {

      return request(app).get('/v1/forecast/&1$7japan?*,jp')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(200)
        .then((res) => {
          expect(res.body).toMatchSnapshot({
            coord: {
              lat: expect.any(Number),
              lon: expect.any(Number),
            },
            forescast: expect.any(Object)
          });

          done();
        });
    }, 30000);

    it('GET /v1/forecast/ol with city name very short', async (done) => {

      return request(app).get('/v1/forecast/ol')
        .set('Content-Type', 'application/json')
        .set('Accept', 'application/json')
        .set('X-Client-IP', '186.122.141.220')
        .expect(400)
        .then((res) => {
          expect(res.body).toMatchSnapshot();

          done();
        });
    }, 30000);
  });
});