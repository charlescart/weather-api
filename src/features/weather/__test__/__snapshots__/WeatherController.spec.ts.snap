// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/current without city name and with IP from Bengaluru/India 1`] = `
Object {
  "city": "Bengaluru",
  "coord": Any<Object>,
  "country": "IN",
  "description": Any<String>,
  "feelsLike": Any<Number>,
  "temp": Any<Number>,
  "tempMax": Any<Number>,
  "tempMin": Any<Number>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/current without city name and with IP from Hurlingham/Argentina 1`] = `
Object {
  "city": "Hurlingham",
  "coord": Any<Object>,
  "country": "AR",
  "description": Any<String>,
  "feelsLike": Any<Number>,
  "temp": Any<Number>,
  "tempMax": Any<Number>,
  "tempMin": Any<Number>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/current/127barcelona52es7329437 with city name impossible 1`] = `
Object {
  "city": "Barcelona",
  "coord": Any<Object>,
  "country": "ES",
  "description": Any<String>,
  "feelsLike": Any<Number>,
  "temp": Any<Number>,
  "tempMax": Any<Number>,
  "tempMin": Any<Number>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/current/caracas with city name Caracas/Venezuela :D :D :D 1`] = `
Object {
  "city": "Caracas",
  "coord": Any<Object>,
  "country": "VE",
  "description": Any<String>,
  "feelsLike": Any<Number>,
  "temp": Any<Number>,
  "tempMax": Any<Number>,
  "tempMin": Any<Number>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/current/h with city name very short 1`] = `
Object {
  "entry": Object {
    "params": Object {},
    "query": Object {},
  },
  "response": Object {
    "error": "BadRequestError",
    "msg": Array [
      Object {
        "constraints": Object {
          "minLength": "city must be longer than or equal to 3 characters",
        },
        "property": "city",
      },
    ],
  },
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/forecast without city name and with IP from Toronto/Canada 1`] = `
Object {
  "climateReferenceZone": "Downtown Toronto",
  "coord": Object {
    "lat": Any<Number>,
    "lon": Any<Number>,
  },
  "country": "CA",
  "forescast": Any<Object>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/forecast/&1$7japan?*,jp with city name impossible 1`] = `
Object {
  "climateReferenceZone": "Japan",
  "coord": Object {
    "lat": Any<Number>,
    "lon": Any<Number>,
  },
  "country": "JP",
  "forescast": Any<Object>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/forecast/japan with city name Japan/Japan 1`] = `
Object {
  "climateReferenceZone": "Japan",
  "coord": Object {
    "lat": Any<Number>,
    "lon": Any<Number>,
  },
  "country": "JP",
  "forescast": Any<Object>,
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/forecast/ol with city name very short 1`] = `
Object {
  "entry": Object {
    "params": Object {},
    "query": Object {},
  },
  "response": Object {
    "error": "BadRequestError",
    "msg": Array [
      Object {
        "constraints": Object {
          "minLength": "city must be longer than or equal to 3 characters",
        },
        "property": "city",
      },
    ],
  },
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/location with IP from San Francisco/US 1`] = `
Object {
  "clientIp": "162.243.152.177",
  "infoFromYourIp": Object {
    "city": "San Francisco",
    "coord": Object {
      "lat": Any<Number>,
      "lon": Any<Number>,
    },
    "country": "US",
  },
  "weather": Object {
    "climateReferenceZone": "San Francisco",
    "coord": Object {
      "lat": Any<Number>,
      "lon": Any<Number>,
    },
    "country": "US",
    "forescast": Any<Object>,
  },
}
`;

exports[`WeatherController EndPoints Location, Current and Forecast GET /v1/location with IP invalid ::1 in localhost 1`] = `
Object {
  "entry": Object {
    "params": Object {},
    "query": Object {},
  },
  "response": Object {
    "error": "BadRequestError",
    "msg": Array [
      Object {
        "constraints": Object {
          "isIp": "clientIp must be an ip address",
        },
        "property": "clientIp",
      },
    ],
  },
}
`;
